File: internal/application/usecase/role/delete.go
package role

import (
	"context"

	"go-clean-template/internal/domain"
)

type DeleteRoleInput struct {
	ID string
}

type DeleteRoleUseCase interface {
	Execute(ctx context.Context, input *DeleteRoleInput) error
}

type deleteRoleUseCase struct {
	roleRepo domain.RoleRepository
}

func NewDeleteRoleUseCase(roleRepo domain.RoleRepository) DeleteRoleUseCase {
	return &deleteRoleUseCase{roleRepo: roleRepo}
}

func (uc *deleteRoleUseCase) Execute(ctx context.Context, input *DeleteRoleInput) error {
	if _, err := uc.roleRepo.FindByID(ctx, input.ID); err != nil {
		return err
	}
	return uc.roleRepo.Delete(ctx, input.ID)
}

File: src/application/usecase/role/delete.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::RoleRepository;

pub struct DeleteRoleInput {
    pub id: String,
}

#[async_trait]
pub trait DeleteRoleUseCase: Send + Sync {
    async fn execute(&self, input: DeleteRoleInput) -> Result<(), AppError>;
}

pub struct DeleteRoleUseCaseImpl {
    role_repo: Arc<dyn RoleRepository>,
}

impl DeleteRoleUseCaseImpl {
    pub fn new(role_repo: Arc<dyn RoleRepository>) -> Self {
        DeleteRoleUseCaseImpl { role_repo }
    }
}

#[async_trait]
impl DeleteRoleUseCase for DeleteRoleUseCaseImpl {
    async fn execute(&self, input: DeleteRoleInput) -> Result<(), AppError> {
        self.role_repo.find_by_id(&input.id).await?;
        self.role_repo.delete(&input.id).await?;
        Ok(())
    }
}
